### PR TITLE
CLI: Dump and replace support

### PR DESCRIPTION
**Pull request by Miepee**
This PR adds the following functionality to the CLI:
- dump and replace individual code entries
- dump and replace all code entries
- dump all embedded textures
- replace all or individual embedded textures
- dump all strings

It does currently not add the functionality to the interactive menu. The only way right now to access these is via the parameters.

### CODE REVIEW

File: UndertaleModCli/Program.cs [hunk 1]
```diff
@@ -270,6 +319,140 @@ private static int Info(InfoOptions options)
...
+                program = new Program(options.Datafile, options.Verbose, options.Output);
+            }
+            catch (FileNotFoundException e)
+            {
+                Console.Error.WriteLine(e.Message);
+                return EXIT_FAILURE;
+            }
+
+            // If user provided code to dump, dump code
+            if ((options.Code != null) && (options.Code.Length > 0) && (program.Data.Code.Count > 0))
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:341]: Can be replaced to `if (options.Code?.Length > 0 && program.Data.Code.Count > 0)`

File: UndertaleModCli/Program.cs [hunk 2]
```diff
@@ -270,6 +319,140 @@ private static int Info(InfoOptions options)
...
+                program = new Program(options.Datafile, null, options.Output, options.Verbose);
+            }
+            catch (FileNotFoundException e)
+            {
+                Console.Error.WriteLine(e.Message);
+                return EXIT_FAILURE;
+            }
+
+            // If user provided code to replace, replace them
+            if ((options.Code != null) && (options.Code.Length > 0) && (program.Data.Code.Count > 0))
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:384]: Can be replaced to `if (options.Code?.Length > 0 && program.Data.Code.Count > 0)`

File: UndertaleModCli/Program.cs [hunk 3]
```diff
@@ -270,6 +319,140 @@ private static int Info(InfoOptions options)
...
+                // Otherwise, just replace every file which was given
+                else
+                {
+                    foreach (KeyValuePair<string, FileInfo> keyValue in codeDict)
+                        program.ReplaceCodeEntryWithFile(keyValue.Key, keyValue.Value);
+                }
+            }
+
+            // If user provided texture to replace, replace them
+            if ((options.Textures != null) && (options.Textures.Length > 0))
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:417]: Can be replaced to `if (options.Textures?.Length > 0)`

File: UndertaleModCli/Program.cs [hunk 4]
```diff
@@ -389,6 +573,111 @@ private void CliQuickInfo()
             if (IsInteractive) Pause();
         }
 
+        /// <summary>
+        /// Dumps a code entry from a data file.
+        /// </summary>
+        /// <param name="codeEntry">The code entry that should get dumped</param>
+        private void DumpCodeEntry(string codeEntry)
+        {
+            UndertaleCode? code = Data.Code.FirstOrDefault(c => c.Name.Content == codeEntry);
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:582]:
```suggestion
            UndertaleCode code = Data.Code.ByName(codeEntry);
```

File: UndertaleModCli/Program.cs [hunk 5]
```diff
@@ -389,6 +573,111 @@ private void CliQuickInfo()
...
+        }
+
+        /// <summary>
+        /// Replaces a code entry with text from another file.
+        /// </summary>
+        /// <param name="codeEntry">The code entry to replace</param>
+        /// <param name="fileToReplace">File path which should replace the code entry.</param>
+        private void ReplaceCodeEntryWithFile(string codeEntry, FileInfo fileToReplace)
+        {
+            UndertaleCode? code = Data.Code.FirstOrDefault(c => c.Name.Content == codeEntry);
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:646]:
```suggestion
            UndertaleCode code = Data.Code.ByName(codeEntry);
```

File: UndertaleModCli/Program.cs [hunk 6]
```diff
@@ -389,6 +573,111 @@ private void CliQuickInfo()
...
+        }
+
+        /// <summary>
+        /// Replaces an embedded texture with contents from another file.
+        /// </summary>
+        /// <param name="textureEntry">Embedded texture to replace</param>
+        /// <param name="fileToReplace">File path which should replace the embedded texture.</param>
+        private void ReplaceTextureWithFile(string textureEntry, FileInfo fileToReplace)
+        {
+            UndertaleEmbeddedTexture? texture = Data.EmbeddedTextures.FirstOrDefault(t => t.Name.Content == textureEntry);
```
Comment (VladiStep) [inline UndertaleModCli/Program.cs:667]:
```suggestion
            UndertaleEmbeddedTexture texture = Data.EmbeddedTextures.ByName(textureEntry);
```

**Comment by suphamster** [comment]
I can't get to work cli mass embed textures import after cli dump command working (using Win10x64).
To reproduce bug: 
1) dump embed textures: "UndertaleModCli dump -t data.win"
2) import 'em back with:  "UndertaleModCli replace -t UMT_REPLACE_ALL=./EmbeddedTextures -o temp.win data.win"
Getting errors like:
Data file does not contain an embedded texture named Texture 0.png!
Data file does not contain an embedded texture named Texture 1.png!
If I'm do import all emb textures in GUI it works (but after reneming "Texture 0.png" -> "0.png" and etc.).

**Comment by Miepee** [comment]
Commenting under a merged PR is the wrong place for this, open a new issue for it please.

**Comment by suphamster** [comment]
> Commenting under a merged PR is the wrong place for this, open a new issue for it please.

Ok, sorry. Plz delete it created new issue there https://github.com/krzys-h/UndertaleModTool/issues/1058

**Pull request merged**